hbt/production/tau.py

```diff
@@ -280,76 +303,99 @@ def trigger_weights(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    # not existing yet
+    # tautauvbf_mask = flat_np_view(default_tautau_mask & tautauvbf_trigger_passed, axis=1)
+    etau_mask = (channel_id == ch_etau.id) & cross_triggered & (ak.local_index(events.Tau) == 0)
+    flat_etau_mask = flat_np_view(etau_mask, axis=1)
 
-    #
-    # compute varied trigger weights
-    #
+    mutau_mask = (channel_id == ch_mutau.id) & cross_triggered & (ak.local_index(events.Tau) == 0)
+    flat_mutau_mask = flat_np_view(mutau_mask, axis=1)
 
-    for direction in ["up", "down"]:
-        for ch, ch_corr, mask in [
-            ("etau", "etau", etau_mask),
-            ("mutau", "mutau", mutau_mask),
-            ("tautau", "ditau", tautau_mask),
-            # ("tautauvbf", "ditauvbf", tautauvbf_mask),
-        ]:
-            sf_unc = sf_nom.copy()
-            sf_unc[mask] = self.trigger_corrector(*eval_args(mask, ch_corr, direction))
-            events = set_ak_column_f32(events, f"tau_trigger_weight_{ch}_{direction}", reduce_mul(sf_unc))
+    # start with flat ones
+    for kind in ["data", "mc"]:
+        wp_config = self.config_inst.x.tau_trigger_working_points
+        eval_args = lambda mask, ch, syst: (pt[mask], dm[mask], ch, wp_config.trigger_corr, f"eff_{kind}", syst)
+        for corr_channel in ["etau", "mutau", "tautau", "tautaujet"]:  # TODO: add tautauvbf
+            if corr_channel == "etau":
+                mask = flat_etau_mask
+                corr_channel_arg = corr_channel
+            elif corr_channel == "mutau":
+                mask = flat_mutau_mask
+                corr_channel_arg = corr_channel
+            elif corr_channel == "tautau":
+                mask = flat_tautau_mask
+                corr_channel_arg = "ditau"
+            elif corr_channel == "tautaujet":
+                mask = flat_tautaujet_mask
+                corr_channel_arg = "ditaujet"
+            else:
+                raise ValueError(f"Unknown channel {corr_channel}")
+            sf_nom = np.ones_like(pt, dtype=np.float32)
+            sf_nom[mask] = self.tau_trig_corrector(*eval_args(mask, corr_channel_arg, "nom"))
+            # create and store weights
+            events = set_ak_column_f32(
+                events,
+                f"tau_trigger_eff_{kind}_{corr_channel}",
+                reduce_mul(sf_nom),
+            )
 
+        #
+        # compute varied trigger weights
+        #
+
+        for direction in ["up", "down"]:
+            for ch, ch_corr, mask in [
+                ("etau", "etau", etau_mask),
+                ("mutau", "mutau", mutau_mask),
+                ("tautau", "ditau", tautau_mask),
+                ("tautaujet", "ditaujet", tautaujet_mask),
+                # ("tautauvbf", "ditauvbf", tautauvbf_mask),
+            ]:
+                for decay_mode in [0, 1, 10, 11]:
+                    decay_mode_mask = mask & (events.Tau.decayMode == decay_mode)
+                    flat_decay_mode_mask = flat_np_view(decay_mode_mask, axis=1)
```

Comment:
Could you make the `direction` loop the inner-most one, starting below `flat_decay_mode_mask = ...`? Then we can half the number of times we create the joined mask.